File: src/main/java/com/portofino/realtrainmodunofficial/client/screen/ModelSelectScreen.java
package com.portofino.realtrainmodunofficial.client.screen;

import com.portofino.realtrainmodunofficial.client.PackButtonTextureCache;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiGraphics;
import net.minecraft.client.gui.components.Button;
import net.minecraft.client.gui.components.EditBox;
import net.minecraft.client.gui.components.ObjectSelectionList;
import net.minecraft.client.gui.screens.Screen;
import net.minecraft.network.chat.Component;
import net.minecraft.util.Mth;
import net.neoforged.api.distmarker.Dist;
import net.neoforged.api.distmarker.OnlyIn;

import java.util.Comparator;
import java.util.List;
import java.util.function.Consumer;

@OnlyIn(Dist.CLIENT)
public class ModelSelectScreen extends Screen {
    public record SelectionResult(String modelId, String dataMapValue) {}

    public record ModelInfo(String id, String displayName, String packName, String buttonTexture, String category) {
        public ModelInfo(String id, String displayName, String packName, String buttonTexture) {
            this(id, displayName, packName, buttonTexture, "");
        }
    }

    private static final int LIST_TOP = 32;
    private static final int LIST_BOTTOM_MARGIN = 16;
    private static final int ITEM_HEIGHT_MIN = 38;
    private static final int ITEM_HEIGHT_MAX = 72;

    private final List<ModelInfo> models;
    private final Consumer<SelectionResult> onSelected;
    private final String initialSelectedId;
    private final String initialDataMapValue;
    private ModelList modelList;
    private EditBox dataMapBox;

    public ModelSelectScreen(Component title, List<ModelInfo> models, Consumer<SelectionResult> onSelected) {
        this(title, models, onSelected, null, "");
    }

    public ModelSelectScreen(Component title, List<ModelInfo> models, Consumer<SelectionResult> onSelected, String initialSelectedId, String initialDataMapValue) {
        super(title);
        this.models = models.stream()
            .sorted(Comparator
                .comparing((ModelInfo info) -> safe(info.category()), String.CASE_INSENSITIVE_ORDER)
                .thenComparing(info -> safe(info.displayName()), String.CASE_INSENSITIVE_ORDER)
                .thenComparing(info -> safe(info.id()), String.CASE_INSENSITIVE_ORDER))
            .toList();
        this.onSelected = onSelected;
        this.initialSelectedId = initialSelectedId;
        this.initialDataMapValue = initialDataMapValue == null ? "" : initialDataMapValue;
    }

    private static String safe(String value) {
        return value == null ? "" : value;
    }

    @Override
    protected void init() {
        int listWidth = Mth.clamp(width / 2 - 28, 220, 360);
        int itemHeight = computeItemHeight(listWidth);
        modelList = new ModelList(minecraft, listWidth, height - LIST_BOTTOM_MARGIN, LIST_TOP, itemHeight);
        addRenderableWidget(modelList);

        int rightLeft = listWidth + 22;
        int rightWidth = Math.max(180, width - rightLeft - 18);
        dataMapBox = new EditBox(font, rightLeft, LIST_TOP + 6, rightWidth, 20, Component.literal("no_name"));
        dataMapBox.setValue(initialDataMapValue);
        addRenderableWidget(dataMapBox);
        setInitialFocus(dataMapBox);

        int buttonWidth = Math.min(150, Math.max(72, (rightWidth - 6) / 2));
        int buttonsLeft = rightLeft;
        int buttonsTop = dataMapBox.getY() + dataMapBox.getHeight() + 8;

        addRenderableWidget(Button.builder(Component.translatable("gui.done"), btn -> {
            ModelList.ModelEntry selected = modelList.getSelected();
            if (selected != null) {
                onSelected.accept(new SelectionResult(selected.id, dataMapBox == null ? "" : dataMapBox.getValue()));
            }
            onClose();
        }).bounds(buttonsLeft, buttonsTop, buttonWidth, 20).build());

        addRenderableWidget(Button.builder(Component.translatable("gui.cancel"), btn -> onClose())
            .bounds(buttonsLeft + buttonWidth + 6, buttonsTop, buttonWidth, 20)
            .build());
    }

    private int computeItemHeight(int listWidth) {
        int itemHeight = ITEM_HEIGHT_MIN;
        for (ModelInfo info : models) {
            PackButtonTextureCache.ButtonTextureInfo buttonTexture = PackButtonTextureCache.get(info.packName(), info.buttonTexture());
            if (buttonTexture == null) {
                continue;
            }
            int sourceHeight = Math.max(1, buttonTexture.sourceHeight());
            itemHeight = Math.max(itemHeight, Mth.clamp(sourceHeight + 4, ITEM_HEIGHT_MIN, ITEM_HEIGHT_MAX));
        }
        return itemHeight;
    }

    @Override
    public void render(GuiGraphics graphics, int mouseX, int mouseY, float partialTick) {
        renderBackground(graphics, mouseX, mouseY, partialTick);
        super.render(graphics, mouseX, mouseY, partialTick);
        graphics.drawCenteredString(font, getTitle(), width / 2, 10, 0xFFFFFF);
        if (dataMapBox != null) {
            graphics.drawString(font, Component.literal("no_name"), dataMapBox.getX(), dataMapBox.getY() - 12, 0xFFFFFF, false);
        }
        if (models.isEmpty()) {
            graphics.drawCenteredString(font,
                Component.translatable("screen.realtrainmodunofficial.no_models"),
                width / 2, height / 2, 0xAAAAAA);
        }
    }

    @Override
    public boolean isPauseScreen() {
        return false;
    }

    private class ModelList extends ObjectSelectionList<ModelList.ModelEntry> {
        ModelList(Minecraft mc, int width, int height, int top, int itemHeight) {
            super(mc, width, height, top, itemHeight);
            setRenderHeader(false, 0);
            String lastCategory = null;
            ModelEntry initialEntry = null;
            int initialRow = -1;
            int row = 0;
            for (ModelInfo info : models) {
                // packName（zipファイル名）を category fallback に使わない。
                // vehicleType が無い項目はヘッダーなしでフラットに並べる。
                String category = info.category();
                if (category != null && !category.isBlank() && !category.equals(lastCategory)) {
                    addEntry(new ModelEntry(category));
                    lastCategory = category;
                    row++;
                }
                ModelEntry entry = new ModelEntry(info.id(), info.displayName(), info.packName(), info.buttonTexture());
                addEntry(entry);
                if (initialSelectedId != null && initialSelectedId.equals(info.id())) {
                    initialEntry = entry;
                    initialRow = row;
                }
                row++;
            }
            if (initialEntry == null) {
                for (ModelEntry entry : children()) {
                    if (!entry.header) {
                        initialEntry = entry;
                        initialRow = 0;
                        break;
                    }
                }
            }
            if (initialEntry != null) {
                setSelected(initialEntry);
                setScrollAmount(Math.max(0.0D, initialRow * (double) itemHeight - height * 0.5D));
            }
        }

        @Override
        protected int getScrollbarPosition() {
            return width - 8;
        }

        @Override
        public int getRowWidth() {
            return width - 20;
        }

        class ModelEntry extends ObjectSelectionList.Entry<ModelEntry> {
            public final String id;
            private final Component label;
            private final PackButtonTextureCache.ButtonTextureInfo buttonTexture;
            private final boolean header;

            ModelEntry(String category) {
                this.id = "";
                this.label = Component.literal(category);
                this.buttonTexture = null;
                this.header = true;
            }

            ModelEntry(String id, String displayName, String packName, String buttonTexturePath) {
                this.id = id;
                String text = (displayName != null && !displayName.isBlank()) ? displayName : id;
                this.label = Component.literal(text);
                this.buttonTexture = PackButtonTextureCache.get(packName, buttonTexturePath);
                this.header = false;
            }

            @Override
            public void render(GuiGraphics graphics, int index, int top, int left, int width,
                               int height, int mouseX, int mouseY, boolean hovered, float partialTick) {
                if (header) {
                    graphics.drawString(ModelSelectScreen.this.font, label, left + 8, top + 14, 0xFFD8D8D8, false);
                    return;
                }
                int previewLeft = left;
                int previewTop = top;
                int previewHeight = height;
                int previewWidth = Math.max(1, width);
                if (buttonTexture != null) {
                    int sourceX = buttonTexture.sourceX();
                    int sourceY = buttonTexture.sourceY();
                    int sourceWidth = Math.max(1, buttonTexture.sourceWidth());
                    int sourceHeight = Math.max(1, buttonTexture.sourceHeight());
                    // Fit within cell: scale to fit both dimensions, maintain aspect ratio
                    float scaleH = (float) previewHeight / sourceHeight;
                    float scaleW = (float) previewWidth / sourceWidth;
                    float scale = Math.min(scaleH, scaleW);
                    int destWidth = Math.max(1, Math.round(sourceWidth * scale));
                    int destHeight = Math.max(1, Math.round(sourceHeight * scale));
                    int offsetX = (previewWidth - destWidth) / 2;
                    int offsetY = (previewHeight - destHeight) / 2;
                    graphics.blit(
                        buttonTexture.location(),
                        previewLeft + offsetX,
                        previewTop + offsetY,
                        destWidth,
                        destHeight,
                        sourceX,
                        sourceY,
                        sourceWidth,
                        sourceHeight,
                        buttonTexture.width(),
                        buttonTexture.height()
                    );
                } else {
                    graphics.drawCenteredString(ModelSelectScreen.this.font, label, previewLeft + (previewWidth / 2), top + (height - 8) / 2, 0xAAAAAA);
                }
            }
            
            @Override
            public boolean mouseClicked(double mouseX, double mouseY, int button) {
                if (header) {
                    return false;
                }
                if (button == 0) {
                    ModelList.this.setSelected(this);
                    return true;
                }
                return false;
            }

            @Override
            public Component getNarration() {
                return label;
            }
        }
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/client/screen/ModelSelectScreen.kt
package cc.mirukuneko.realtrainmodrenewed.client.screen

import cc.mirukuneko.realtrainmodrenewed.client.PackButtonTextureCache
import cc.mirukuneko.realtrainmodrenewed.client.model.MqoModelLoader
import cc.mirukuneko.realtrainmodrenewed.client.renderer.BogieRenderer
import cc.mirukuneko.realtrainmodrenewed.entity.TrainEntity
import cc.mirukuneko.realtrainmodrenewed.installedobject.InstalledObjectRegistry
import cc.mirukuneko.realtrainmodrenewed.rail.RailRegistry
import cc.mirukuneko.realtrainmodrenewed.vehicle.VehicleDefinition
import cc.mirukuneko.realtrainmodrenewed.vehicle.VehicleRegistry
import com.mojang.blaze3d.vertex.PoseStack
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.GuiGraphicsExtractor
import net.minecraft.client.gui.components.Button
import net.minecraft.client.gui.components.EditBox
import net.minecraft.client.gui.components.ObjectSelectionList
import net.minecraft.client.gui.screens.Screen
import net.minecraft.client.input.MouseButtonEvent
import net.minecraft.client.renderer.MultiBufferSource
import net.minecraft.client.renderer.RenderPipelines
import net.minecraft.client.renderer.texture.OverlayTexture
import net.minecraft.network.chat.Component
import net.minecraft.util.LightCoordsUtil
import net.minecraft.util.Mth
import net.minecraft.world.phys.Vec3
import java.lang.String.CASE_INSENSITIVE_ORDER
import java.util.Locale
import java.util.concurrent.ConcurrentHashMap
import java.util.function.Consumer
import kotlin.math.max
import kotlin.math.min

open class ModelSelectScreen @JvmOverloads constructor(
    title: Component,
    models: List<ModelInfo>,
    private val onSelected: Consumer<SelectionResult>,
    private val initialSelectedId: String? = null,
    initialDataMapValue: String? = "",
) : Screen(title) {
    @JvmRecord
    data class SelectionResult(val modelId: String?, val dataMapValue: String?)

    @JvmRecord
    data class ModelInfo @JvmOverloads constructor(
        val id: String?,
        val displayName: String?,
        val packName: String?,
        val buttonTexture: String?,
        val category: String? = "",
    )

    private val models: List<ModelInfo> = models.sortedWith(
        compareBy<ModelInfo, String>(CASE_INSENSITIVE_ORDER) { safe(it.category) }
            .thenBy(CASE_INSENSITIVE_ORDER) { safe(it.displayName) }
            .thenBy(CASE_INSENSITIVE_ORDER) { safe(it.id) },
    )
    private val initialDataMapValue: String = initialDataMapValue ?: ""

    private lateinit var modelList: ModelList
    private var dataMapBox: EditBox? = null
    private var selectedId: String? = null

    private var previewYaw = 0.0f
    private var previewPitch = 15.0f
    private var previewUserRotated = false
    private var previewDragging = false
    private var lastDragX = 0.0
    private var lastDragY = 0.0
    private var previewZoom = 1.0f
    private var previewEntity: TrainEntity? = null
    private var previewEntityId: String? = null

    private fun listWidth(): Int = BTN_W + 16
    private fun rightLeft(): Int = listWidth() + 4
    private fun rightWidth(): Int = max(100, width - rightLeft() - 4)
    private fun previewSize(): Int = min(rightWidth(), height - LIST_TOP - 60)

    private fun fitText(text: String?, maxWidth: Int): String {
        if (text.isNullOrBlank() || font.width(text) <= maxWidth) {
            return text ?: ""
        }
        val suffix = "..."
        val suffixWidth = font.width(suffix)
        if (maxWidth <= suffixWidth) {
            return font.plainSubstrByWidth(text, max(0, maxWidth))
        }
        return font.plainSubstrByWidth(text, maxWidth - suffixWidth) + suffix
    }

    private fun isInPreviewArea(mx: Double, my: Double): Boolean {
        val rl = rightLeft()
        val rw = rightWidth()
        val ps = previewSize()
        return mx >= rl && mx <= rl + rw && my >= LIST_TOP && my <= LIST_TOP + ps
    }

    override fun mouseClicked(event: MouseButtonEvent, doubleClick: Boolean): Boolean {
        val mx = event.x()
        val my = event.y()
        if (event.button() == 0 && isInPreviewArea(mx, my)) {
            previewDragging = true
            lastDragX = mx
            lastDragY = my
            return true
        }
        return super.mouseClicked(event, doubleClick)
    }

    override fun mouseDragged(event: MouseButtonEvent, dragX: Double, dragY: Double): Boolean {
        val mx = event.x()
        val my = event.y()
        if (previewDragging && event.button() == 0) {
            previewUserRotated = true
            previewYaw += (mx - lastDragX).toFloat() * 0.8f
            previewPitch += (my - lastDragY).toFloat() * 0.8f
            previewPitch = Mth.clamp(previewPitch, -89.0f, 89.0f)
            lastDragX = mx
            lastDragY = my
            return true
        }
        return super.mouseDragged(event, dragX, dragY)
    }

    override fun mouseReleased(event: MouseButtonEvent): Boolean {
        if (event.button() == 0 && previewDragging) {
            previewDragging = false
            return true
        }
        return super.mouseReleased(event)
    }

    override fun mouseScrolled(mx: Double, my: Double, scrollX: Double, scrollY: Double): Boolean {
        if (isInPreviewArea(mx, my)) {
            previewZoom = Mth.clamp(previewZoom * if (scrollY > 0) 1.1f else 1.0f / 1.1f, 0.3f, 6.0f)
            return true
        }
        return super.mouseScrolled(mx, my, scrollX, scrollY)
    }

    override fun init() {
        val lw = listWidth()
        val listHeight = height - LIST_TOP - LIST_BOTTOM_MARGIN
        modelList = ModelList(minecraft, lw, listHeight, LIST_TOP, BTN_H)
        addRenderableWidget(modelList)

        val rl = rightLeft()
        val rw = rightWidth()
        val datamapY = LIST_TOP + previewSize() + 8
        val dataMap = EditBox(font, rl, datamapY, rw, 20, Component.empty())
        dataMap.value = initialDataMapValue
        dataMapBox = dataMap
        addRenderableWidget(dataMap)

        val buttonWidth = min(100, max(60, (rw - 4) / 2))
        val buttonY = datamapY + 28
        addRenderableWidget(
            Button.builder(Component.translatable("gui.done")) {
                val selected = modelList.selected
                if (selected != null && !selected.header) {
                    onSelected.accept(SelectionResult(selected.id, dataMapBox?.value ?: ""))
                }
                onClose()
            }.bounds(rl, buttonY, buttonWidth, 20).build(),
        )
        addRenderableWidget(
            Button.builder(Component.translatable("gui.cancel")) { onClose() }
                .bounds(rl + buttonWidth + 4, buttonY, buttonWidth, 20)
                .build(),
        )
    }

    override fun extractRenderState(graphics: GuiGraphicsExtractor, mouseX: Int, mouseY: Int, partialTick: Float) {
        super.extractRenderState(graphics, mouseX, mouseY, partialTick)
        graphics.centeredText(font, title, listWidth() / 2, 7, 0xFFFFFF)

        val rl = rightLeft()
        val rw = rightWidth()
        val ps = previewSize()
        val previewLeft = rl
        val previewTop = LIST_TOP

        graphics.fill(previewLeft, previewTop, previewLeft + rw, previewTop + ps, -0x78000000)

        val selected = modelList.selected
        if (selected != null && !selected.header) {
            renderPreviewPanel(graphics, previewLeft, previewTop, rw, ps, selected)
        }

        if (models.isEmpty()) {
            graphics.centeredText(
                font,
                Component.translatable("screen.realtrainmodrenewed.no_models"),
                previewLeft + rw / 2,
                previewTop + ps / 2,
                0xAAAAAA,
            )
        }
    }

    private fun renderPreviewPanel(graphics: GuiGraphicsExtractor, left: Int, top: Int, width: Int, height: Int, entry: ModelList.ModelEntry) {
        val name = entry.label.string
        graphics.text(font, Component.literal(fitText(name, width - 12)), left + 6, top + 6, -0x1, false)

        val infoTop = top + height - 31
        graphics.fill(left, infoTop - 4, left + width, top + height, -0x56000000)
        graphics.text(font, Component.literal(fitText(entry.packName, width - 12)), left + 6, infoTop, -0x474738, false)
        graphics.text(font, Component.literal(fitText(entry.id, width - 12)), left + 6, infoTop + 11, -0x777760, false)

        val imageTop = top + 24
        val imageHeight = max(20, infoTop - imageTop - 8)
        if (entry.buttonTex != null) {
            val drawW = min(width - 24, BTN_W * 2)
            val drawH = max(BTN_H, min(imageHeight, Math.round(drawW * (BTN_H / BTN_W.toFloat()))))
            val drawX = left + (width - drawW) / 2
            val drawY = imageTop + max(0, (imageHeight - drawH) / 2)
            graphics.fill(drawX - 2, drawY - 2, drawX + drawW + 2, drawY + drawH + 2, -0xefefe8)
            graphics.blit(
                RenderPipelines.GUI_TEXTURED,
                entry.buttonTex.location,
                drawX,
                drawY,
                entry.buttonTex.sourceX.toFloat(),
                entry.buttonTex.sourceY.toFloat(),
                drawW,
                drawH,
                entry.buttonTex.sourceWidth,
                entry.buttonTex.sourceHeight,
                entry.buttonTex.width,
                entry.buttonTex.height,
            )
        } else {
            val boxW = min(width - 24, 220)
            val boxH = min(imageHeight, 64)
            val boxX = left + (width - boxW) / 2
            val boxY = imageTop + max(0, (imageHeight - boxH) / 2)
            graphics.fill(boxX, boxY, boxX + boxW, boxY + boxH, -0xeeee8)
            graphics.fill(boxX, boxY, boxX + boxW, boxY + 1, -0xa5a590)
            graphics.fill(boxX, boxY + boxH - 1, boxX + boxW, boxY + boxH, -0xa5a590)
            graphics.fill(boxX, boxY, boxX + 1, boxY + boxH, -0xa5a590)
            graphics.fill(boxX + boxW - 1, boxY, boxX + boxW, boxY + boxH, -0xa5a590)
            graphics.centeredText(font, Component.literal("No preview image"), left + width / 2, imageTop + imageHeight / 2 - 4, -0x555556)
        }
    }

    private fun getOrLoadModel(id: String?, packName: String?): MqoModelLoader.MqoModel? {
        if (id.isNullOrBlank() || MISSING_MODEL_CACHE.contains(id)) return null
        if (MODEL_CACHE.containsKey(id)) return MODEL_CACHE[id]
        var model: MqoModelLoader.MqoModel? = null
        try {
            val vehicleDefinition = VehicleRegistry.getById(id)
            if (vehicleDefinition != null && vehicleDefinition.getModelFile() != null && vehicleDefinition.getModelFile().isNotBlank()) {
                model = MqoModelLoader.loadModelForVehicle(vehicleDefinition)
            }
            if (model == null) {
                val installedObjectDefinition = InstalledObjectRegistry.getById(id)
                if (installedObjectDefinition != null && installedObjectDefinition.modelFile != null && installedObjectDefinition.modelFile.isNotBlank()) {
                    model = MqoModelLoader.loadModelFromPack(
                        installedObjectDefinition.packName,
                        installedObjectDefinition.modelFile,
                        installedObjectDefinition.textureOverrides,
                        null,
                        installedObjectDefinition.isSmoothing(),
                    )
                }
            }
            if (model == null) {
                val railDefinition = RailRegistry.getById(id)
                if (railDefinition != null && railDefinition.modelFile != null && railDefinition.modelFile.isNotBlank()) {
                    model = MqoModelLoader.loadModelFromPack(
                        railDefinition.packName,
                        railDefinition.modelFile,
                        railDefinition.textureOverrides,
                        null,
                        false,
                    )
                }
            }
        } catch (ignored: Exception) {
        }
        if (model != null) {
            MODEL_CACHE[id] = model
        } else {
            MISSING_MODEL_CACHE.add(id)
        }
        return model
    }

    private fun getOrCreatePreviewEntity(definition: VehicleDefinition?, model: MqoModelLoader.MqoModel): TrainEntity? {
        if (definition == null || definition.id == null) {
            return null
        }
        if (previewEntity != null && definition.id == previewEntityId) {
            return previewEntity
        }
        try {
            val minecraft = Minecraft.getInstance()
            val level = minecraft.level
            if (level == null) {
                return null
            }
            val entity = TrainEntity.create(level, definition.id, 0.0, 0.0, 0.0, 0.0f, definition.trainDistance)
                ?: return null
            if (model.getScriptEngine() != null) {
                entity.scriptEngine = model.getScriptEngine()
            }
            previewEntity = entity
            previewEntityId = definition.id
            return entity
        } catch (throwable: Throwable) {
            return null
        }
    }

    override fun isPauseScreen(): Boolean = false

    private inner class ModelList(
        minecraft: Minecraft,
        width: Int,
        height: Int,
        top: Int,
        itemHeight: Int,
    ) : ObjectSelectionList<ModelList.ModelEntry>(minecraft, width, height, top, itemHeight) {
        init {
            var lastCategory: String? = null
            var initialEntry: ModelEntry? = null
            var initialRow = -1
            var row = 0
            for (info in models) {
                val category = info.category
                if (!category.isNullOrBlank() && category != lastCategory) {
                    addEntry(ModelEntry(category))
                    lastCategory = category
                    row++
                }
                val entry = ModelEntry(info.id, info.displayName, info.packName, info.buttonTexture)
                addEntry(entry)
                if (initialSelectedId != null && initialSelectedId == info.id) {
                    initialEntry = entry
                    initialRow = row
                }
                row++
            }
            if (initialEntry == null) {
                for (entry in children()) {
                    if (!entry.header) {
                        initialEntry = entry
                        initialRow = 0
                        break
                    }
                }
            }
            if (initialEntry != null) {
                selected = initialEntry
                selectedId = initialEntry.id
                setScrollAmount(max(0.0, initialRow * itemHeight.toDouble() - height * 0.5))
            }
        }

        override fun scrollBarX(): Int = width - 6

        override fun getRowWidth(): Int = width - 8

        override fun extractSelection(graphics: GuiGraphicsExtractor, entry: ModelEntry, color: Int) {
        }

        inner class ModelEntry : Entry<ModelEntry> {
            val id: String
            val packName: String
            val label: Component
            val buttonTex: PackButtonTextureCache.ButtonTextureInfo?
            val header: Boolean

            constructor(category: String) {
                id = ""
                packName = ""
                label = Component.literal(category)
                buttonTex = null
                header = true
            }

            constructor(id: String?, displayName: String?, packName: String?, buttonTexturePath: String?) {
                this.id = id ?: ""
                this.packName = packName ?: ""
                label = Component.literal(if (safe(displayName).isBlank()) this.id else safe(displayName))
                buttonTex = PackButtonTextureCache.get(this.packName, buttonTexturePath, this.id, displayName)
                header = false
            }

            override fun extractContent(graphics: GuiGraphicsExtractor, mouseX: Int, mouseY: Int, hovered: Boolean, partialTick: Float) {
                val left = x
                val top = y
                if (header) {
                    graphics.text(this@ModelSelectScreen.font, Component.literal(fitText(label.string, BTN_W - 8)), left + 4, top + 11, -0x2f2f30, false)
                    return
                }
                val selected = this === modelList.selected
                if (selected) {
                    graphics.fill(left, top, left + BTN_W, top + BTN_H, 0x663C7DFF)
                }
                if (buttonTex != null) {
                    graphics.fill(left, top, left + BTN_W, top + BTN_H, -0xe5e5d2)
                    graphics.blit(
                        RenderPipelines.GUI_TEXTURED,
                        buttonTex.location,
                        left,
                        top,
                        buttonTex.sourceX.toFloat(),
                        buttonTex.sourceY.toFloat(),
                        BTN_W,
                        BTN_H,
                        buttonTex.sourceWidth,
                        buttonTex.sourceHeight,
                        buttonTex.width,
                        buttonTex.height,
                    )
                }
                val visibleLabel = fitText(label.string, BTN_W - 8)
                val labelY = top + BTN_H - 11
                graphics.fill(left, labelY - 2, left + BTN_W, top + BTN_H, -0x56000000)
                graphics.text(this@ModelSelectScreen.font, Component.literal(visibleLabel), left + 4, labelY, if (buttonTex != null) -0x1 else -0x555556, false)
                if (selected) {
                    graphics.fill(left, top, left + BTN_W, top + 1, -0x1)
                    graphics.fill(left, top + BTN_H - 1, left + BTN_W, top + BTN_H, -0x1)
                    graphics.fill(left, top, left + 1, top + BTN_H, -0x1)
                    graphics.fill(left + BTN_W - 1, top, left + BTN_W, top + BTN_H, -0x1)
                    val selectedText = Component.translatable("screen.realtrainmodrenewed.selected").string
                    val selectedWidth = this@ModelSelectScreen.font.width(selectedText)
                    graphics.fill(left + BTN_W - selectedWidth - 8, top + 2, left + BTN_W - 2, top + 12, -0x34000000)
                    graphics.text(this@ModelSelectScreen.font, Component.literal(selectedText), left + BTN_W - selectedWidth - 5, top + 3, -0x1, false)
                }
            }

            override fun mouseClicked(event: MouseButtonEvent, doubleClick: Boolean): Boolean {
                if (header) return false
                if (event.button() == 0) {
                    modelList.selected = this
                    selectedId = id
                    return true
                }
                return false
            }

            override fun getNarration(): Component = label
        }
    }

    private companion object {
        private const val BTN_W = 160
        private const val BTN_H = 32
        private const val LIST_TOP = 24
        private const val LIST_BOTTOM_MARGIN = 4

        private val MODEL_CACHE = ConcurrentHashMap<String, MqoModelLoader.MqoModel>()
        private val MISSING_MODEL_CACHE = ConcurrentHashMap.newKeySet<String>()

        private fun safe(value: String?): String = value ?: ""

        private fun computePreviewBounds(model: MqoModelLoader.MqoModel, vehicleDef: VehicleDefinition?): FloatArray {
            val bounds = model.computeBounds()
            if (vehicleDef == null) {
                return bounds
            }

            val scale = vehicleDef.getModelScale()
            val offset = vehicleDef.getModelOffset()
            var minX = (bounds[0] * scale + offset.x).toFloat()
            var minY = (bounds[1] * scale + offset.y).toFloat()
            var minZ = (bounds[2] * scale + offset.z).toFloat()
            var maxX = (bounds[3] * scale + offset.x).toFloat()
            var maxY = (bounds[4] * scale + offset.y).toFloat()
            var maxZ = (bounds[5] * scale + offset.z).toFloat()

            for (bogie in vehicleDef.getBogies()) {
                if (bogie == null || bogie.position() == null) {
                    continue
                }
                val pos = bogie.position()
                val x = (pos.x * scale + offset.x).toFloat()
                val y = ((pos.y + 0.24) * scale + offset.y).toFloat()
                val z = (pos.z * scale + offset.z).toFloat()
                minX = min(minX, x - 1.0f * scale)
                minY = min(minY, y - 1.0f * scale)
                minZ = min(minZ, z - 1.0f * scale)
                maxX = max(maxX, x + 1.0f * scale)
                maxY = max(maxY, y + 1.0f * scale)
                maxZ = max(maxZ, z + 1.0f * scale)
            }

            return floatArrayOf(minX, minY, minZ, maxX, maxY, maxZ)
        }

        private fun renderStablePreviewModel(
            model: MqoModelLoader.MqoModel,
            vehicleDef: VehicleDefinition?,
            poseStack: PoseStack,
            buffer: MultiBufferSource.BufferSource,
            previewEnt: Any?,
        ) {
            poseStack.pushPose()
            try {
                if (vehicleDef != null) {
                    val offset = vehicleDef.getModelOffset()
                    poseStack.translate(offset.x, offset.y, offset.z)
                    val modelScale = vehicleDef.getModelScale()
                    poseStack.scale(modelScale, modelScale, modelScale)
                }

                val previewFilter = MqoModelLoader.GroupPredicate(::shouldRenderPreviewGroup)
                var rendered = false
                if (vehicleDef != null && model.hasRenderScript()) {
                    try {
                        MqoModelLoader.renderModel(model, poseStack, buffer, LightCoordsUtil.FULL_BRIGHT, previewFilter, null, previewEnt)
                        rendered = true
                    } catch (ignored: Throwable) {
                        rendered = false
                    }
                }
                if (!rendered) {
                    MqoModelLoader.renderModelWithoutScript(model, poseStack, buffer, LightCoordsUtil.FULL_BRIGHT, OverlayTexture.NO_OVERLAY, false, previewFilter, null)
                    MqoModelLoader.renderModelWithoutScript(model, poseStack, buffer, LightCoordsUtil.FULL_BRIGHT, OverlayTexture.NO_OVERLAY, true, previewFilter, null)
                }

                if (vehicleDef != null) {
                    val selfDrawsRunningGear = model.hasOwnWheelGroups()
                    val bogies = vehicleDef.getBogies()
                    for (i in bogies.indices) {
                        val bogieDef = bogies[i]
                        if (shouldSkipPreviewBogie(selfDrawsRunningGear, bogieDef)) {
                            continue
                        }
                        try {
                            BogieRenderer.renderBogie(poseStack, i, bogieDef, vehicleDef, null, buffer, LightCoordsUtil.FULL_BRIGHT, 0.0f, 1.0f)
                        } catch (ignored: Throwable) {
                        }
                    }
                }
            } finally {
                poseStack.popPose()
            }
        }

        private fun shouldSkipPreviewBogie(selfDrawsRunningGear: Boolean, bogieDef: VehicleDefinition.BogieDefinition?): Boolean {
            if (bogieDef == null || bogieDef.modelFile() == null || bogieDef.modelFile().isBlank()) {
                return true
            }
            if (BogieRenderer.isDummyBogieModel(bogieDef.modelFile())) {
                return true
            }
            return selfDrawsRunningGear && bogieDef.modelFile().lowercase(Locale.ROOT).endsWith(".class")
        }

        private fun shouldRenderPreviewGroup(groupName: String?): Boolean {
            if (groupName.isNullOrBlank()) {
                return true
            }
            val lower = groupName.trim().lowercase(Locale.ROOT)
            return lower != "shadow" &&
                lower != "_shadow" &&
                lower != "shadowplane" &&
                lower != "hitbox" &&
                lower != "collision" &&
                lower != "collider"
        }
    }
}

